Checking how the experience and age affects the salary along with the collinearity between the independent features.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import statsmodels.api as sm

In [2]:
df=pd.read_csv('experience_salary.csv')

In [3]:
df.head()

YearsExperience   Age  Salary
0              1.1  21.0   39343
1              1.3  21.5   46205
2              1.5  21.7   37731
3              2.0  22.0   43525
4              2.2  22.2   39891

In [5]:
df.shape

(30, 3)

In [11]:
x=df[['YearsExperience','Age']]
y=df['Salary']

In [12]:
x=sm.add_constant(x)

In [13]:
x

const  YearsExperience   Age
0     1.0              1.1  21.0
1     1.0              1.3  21.5
2     1.0              1.5  21.7
3     1.0              2.0  22.0
4     1.0              2.2  22.2
5     1.0              2.9  23.0
6     1.0              3.0  23.0
7     1.0              3.2  23.3
8     1.0              3.2  23.3
9     1.0              3.7  23.6
10    1.0              3.9  23.9
11    1.0              4.0  24.0
12    1.0              4.0  24.0
13    1.0              4.1  24.0
14    1.0              4.5  25.0
15    1.0              4.9  25.0
16    1.0              5.1  26.0
17    1.0              5.3  27.0
18    1.0              5.9  28.0
19    1.0              6.0  29.0
20    1.0              6.8  30.0
21    1.0              7.1  30.0
22    1.0              7.9  31.0
23    1.0              8.2  32.0
24    1.0              8.7  33.0
25    1.0              9.0  34.0
26    1.0              9.5  35.0
27    1.0              9.6  36.0
28    1.0             10.3  37.0
29    1.0             10.5  38.0

In [14]:
#Using OLS technique to fit the data
model=sm.OLS(y,x).fit()

In [15]:
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 Salary   R-squared:                       0.960
Model:                            OLS   Adj. R-squared:                  0.957
Method:                 Least Squares   F-statistic:                     323.9
Date:                Sun, 28 May 2023   Prob (F-statistic):           1.35e-19
Time:                        10:13:54   Log-Likelihood:                -300.35
No. Observations:                  30   AIC:                             606.7
Df Residuals:                      27   BIC:                             610.9
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
===================================================================================
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
const           -6661.9872   2.28e+04     -0.292      0.773   -5.35e+04    4.02e+04
YearsExperience  6153.3533   2337.092      2.633      0.014    1358.037    1.09e+04
Age              1836.0136   1285.034      1.429      0.165    -800.659    4472.686
==============================================================================
Omnibus:                        2.695   Durbin-Watson:                   1.711
Prob(Omnibus):                  0.260   Jarque-Bera (JB):                1.975
Skew:                           0.456   Prob(JB):                        0.372
Kurtosis:                       2.135   Cond. No.                         626.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [17]:
x.iloc[:,1:].corr()

YearsExperience       Age
YearsExperience         1.000000  0.987258
Age                     0.987258  1.000000

Thus we can see that the correlation between years of experience and age is 98%. Also, the P-score for age is > 0.05.  
This tells are that we can drop age as a feature because 98% of it will already be covered by years of experience.

This implies there is multicollinearity in this dataset.